src/zql/ivm/graph/DifferenceStream.ts

```diff
@@ -0,0 +1,92 @@
+import {Multiset} from '../Multiset.js';
+import {Version} from '../types.js';
+import {DifferenceStreamWriter} from './DifferenceStreamWriter.js';
+import {IDifferenceStream} from './IDifferenceStream.js';
+import {LinearCountOperator} from './operators/CountOperator.js';
+import {DifferenceEffectOperator} from './operators/DifferenceEffectOperator.js';
+import {EffectOperator} from './operators/EffectOperator.js';
+import {FilterOperator} from './operators/FilterOperator.js';
+import {MapOperator} from './operators/MapOperator.js';
+
+/**
+ * Used to build up a computation pipeline against a stream and then materialize it.
+ * (Note: materialization of the stream is not yet implemented).
+ *
+ * Encapsulates all the details of wiring together operators, readers, and writers.
+ */
+export class DifferenceStream<T> implements IDifferenceStream<T> {
+  readonly #writer: DifferenceStreamWriter<T>;
+
+  constructor() {
+    this.#writer = new DifferenceStreamWriter<T>();
+  }
+
+  newStream<X>(): DifferenceStream<X> {
+    return new DifferenceStream();
+  }
+
+  map<O>(f: (value: T) => O): DifferenceStream<O> {
+    const ret = this.newStream<O>();
+    new MapOperator<T, O>(this.#writer.newReader(), ret.#writer, f);
```

Comment:
its an unfortunate cycle.

1. `Operator` needs a reference to its inputs. `Operator(...readers)`
2. The inputs need a reference to the operator

Operator sets the operator on the readers during its constructor.

```ts
new Operator(readers);

class Operator {
  constructor(inputs) {
    for (input of inputs) {
      input.setOperator(this);
    }
  }
}
```

We could refactor this to:

```ts
const reader = this.#writer.newReader();
const operator = new MapOperator<T, O>(reader, ret.#writer, f);
reader.setOperator(operator);
```

It does look nicer (and it does ensure references to things do not escape before construction is complete) but then we'd have to remember to call `setOperator` anywhere we create an operator.

---

One other option would be for `Operator` not to have a reference to its input. The inputs would instead call the operator with what it needs to use to compute.

This would create a weird "waitingForData" pattern inside operators that take many inputs though.

```ts
class Operator {
  private pendingInputs = [];
  constructor(private numInputs: number) {}
  inputReceived(inputLine: number, input) {
   pendingInputs[inputLine] = input;
   if (allInputsReceived()) { this.run() }
  }
}
```